File: backend/app/Services/TableStatusDeriver.php
<?php

namespace App\Services;

use App\Enums\DiningSessionStatus;
use App\Enums\OrderStatus;
use App\Enums\PaymentStatus;
use App\Models\DiningSession;
use App\Models\RestaurantTable;

class TableStatusDeriver
{
    public function derive(RestaurantTable $table): string
    {
        $session = $table->relationLoaded('diningSessions')
            ? $table->diningSessions
                ->where('status', DiningSessionStatus::Open)
                ->sortByDesc('id')
                ->first()
            : $table->openSession()->with(['orders', 'payment'])->first();

        if (! $session instanceof DiningSession) {
            return 'available';
        }

        $session->loadMissing(['orders', 'payment']);

        $paymentStatus = $session->payment?->status;

        if ($paymentStatus === PaymentStatus::Paid) {
            return 'paid';
        }

        if ($paymentStatus === PaymentStatus::PaymentPending) {
            return 'payment_pending';
        }

        $orders = $session->orders->filter(
            fn ($order) => $order->status !== OrderStatus::Cancelled
        );

        if ($orders->isEmpty()) {
            return 'occupied';
        }

        $priority = [
            OrderStatus::Pending->value => 1,
            OrderStatus::Accepted->value => 2,
            OrderStatus::Preparing->value => 3,
            OrderStatus::Ready->value => 4,
            OrderStatus::Served->value => 5,
            OrderStatus::Completed->value => 6,
        ];

        $active = $orders
            ->reject(fn ($order) => $order->status === OrderStatus::Completed)
            ->sortByDesc(fn ($order) => $priority[$order->status->value] ?? 0)
            ->first();

        if ($active === null) {
            return 'awaiting_payment';
        }

        return match ($active->status) {
            OrderStatus::Pending => 'pending_orders',
            OrderStatus::Accepted => 'accepted',
            OrderStatus::Preparing => 'preparing',
            OrderStatus::Ready => 'ready',
            OrderStatus::Served => 'served',
            default => 'occupied',
        };
    }
}

File: lib/features/tables/domain/table_status_deriver.dart
import '../../orders/domain/entities/order.dart';
import '../../payments/domain/entities/payment.dart';
import '../../shared/domain/enums.dart';
import '../../customer_session/domain/entities/dining_session.dart';

/// Derives table UI status from session + orders + payment (source of truth).
abstract final class TableStatusDeriver {
  static TableDisplayStatus derive({
    DiningSession? session,
    List<Order> orders = const [],
    Payment? payment,
  }) {
    if (session == null || session.status == DiningSessionStatus.closed) {
      return TableDisplayStatus.available;
    }

    if (payment?.status == PaymentStatus.paid) {
      return TableDisplayStatus.paid;
    }

    final activeOrders =
        orders.where((o) => o.status != OrderStatus.cancelled).toList();

    if (activeOrders.isEmpty) {
      return TableDisplayStatus.occupied;
    }

    if (payment?.status == PaymentStatus.paymentPending ||
        _allFoodDone(activeOrders)) {
      if (_allFoodDone(activeOrders) &&
          (payment == null || payment.status != PaymentStatus.paid)) {
        return TableDisplayStatus.paymentPending;
      }
    }

    if (activeOrders.any((o) => o.status == OrderStatus.ready)) {
      return TableDisplayStatus.ready;
    }
    if (activeOrders.any((o) => o.status == OrderStatus.preparing)) {
      return TableDisplayStatus.preparing;
    }
    if (activeOrders.any(
      (o) =>
          o.status == OrderStatus.pending || o.status == OrderStatus.accepted,
    )) {
      return TableDisplayStatus.orderPlaced;
    }
    if (activeOrders.any((o) => o.status == OrderStatus.served) ||
        activeOrders.every((o) => o.status == OrderStatus.completed)) {
      return TableDisplayStatus.paymentPending;
    }

    return TableDisplayStatus.occupied;
  }

  static bool _allFoodDone(List<Order> orders) {
    if (orders.isEmpty) return false;
    return orders.every(
      (o) =>
          o.status == OrderStatus.served ||
          o.status == OrderStatus.completed,
    );
  }

  static bool canBecomeAvailable({
    DiningSession? session,
    Payment? payment,
    List<Order> orders = const [],
  }) {
    if (session == null || session.status == DiningSessionStatus.closed) {
      return true;
    }
    final unpaidActive = orders.any((o) => !o.isTerminal) ||
        (payment != null &&
            payment.status != PaymentStatus.paid &&
            payment.amountDue > 0);
    return !unpaidActive && payment?.status == PaymentStatus.paid;
  }
}
